packages/commonwealth/client/scripts/hooks/api/useTopics.tsx

```diff
@@ -0,0 +1,294 @@
+import React, { useState } from 'react';
+import { Topic } from 'models';
+import { useMutation, useQueryClient, useQuery } from '@tanstack/react-query';
+import $ from 'jquery';
+import app from 'state';
+import queryKeys from './queryKeys';
+
+const useTopics = () => {
+  // auth state to come from zustand, localstorage is just for demo purposes
+  const JWT = localStorage.getItem('jwt');
+
+  const queryClient = useQueryClient();
+  const [topicsForChainId, setTopicsForChainId] = useState<string>();
+
+  const fetchTopicsQuery = useQuery({
+    queryKey: [queryKeys.TOPICS_FOR_CHAIN(topicsForChainId)],
+    queryFn: async () => {
+      return queryClient.getQueryData([
+        queryKeys.TOPICS_FOR_CHAIN(topicsForChainId),
+      ]);
+    },
+    enabled: !!topicsForChainId,
+    staleTime: 60_000, // cache for 60 seconds
+  });
+
+  const addTopicQuery = useMutation({
+    onMutate: async ({
+      name,
+      description,
+      telegram,
+      featuredInSidebar,
+      featuredInNewPost,
+      tokenThreshold,
+      defaultOffchainTemplate,
+    }: {
+      name: string;
+      description: string;
+      telegram: string;
+      featuredInSidebar: boolean;
+      featuredInNewPost: boolean;
+      tokenThreshold: string;
+      defaultOffchainTemplate: string;
+    }) => {
+      // Cancel any outgoing refetches ,(so they don't overwrite
+      // our optimistic update)
+      await queryClient.cancelQueries({
+        queryKey: [queryKeys.TOPICS_FOR_CHAIN(topicsForChainId)],
+      });
+
+      // fetch old topics from react query store
+      const oldTopics = queryClient.getQueryData([
+        queryKeys.TOPICS_FOR_CHAIN(topicsForChainId),
+      ]);
+
+      // update optimistic value
+      queryClient.setQueryData(
+        [queryKeys.TOPICS_FOR_CHAIN(topicsForChainId)],
+        (t: Topic[]) => {
+          const updatedTopics = [...t];
+
+          const newTopic = new Topic({
+            id: -1, // -1 to indicate a new thread
+            name,
+            description,
+            telegram,
+            featured_in_new_post: featuredInNewPost,
+            featured_in_sidebar: featuredInSidebar,
+            token_threshold: tokenThreshold,
+            default_offchain_template: defaultOffchainTemplate,
+          });
+
+          return [...updatedTopics, newTopic];
+        }
+      );
```

Comment:
this is how we would update optimistic value

Reply:
I am totally in favour of option c, that is ideally the best way to reduce boilerplate here. 
About optimistic updates, they can be done incrementally, yes.

>  if the user was expecting immediate reaction but the API call took too much time

I would say the state updated/revert feels immediate (it updates state before api call - snappy). Did you get a chance to try this?

Reply:
I did not try but saying "from my experience" I meant my previous projects where I had to implement optimistic update.

And sorry for a, b, c, d => it should be 1, 2, 3, 4 as order of execution, not an one-to-pick-option

Reply:
I generally agree that we should optimize as much as possible without introducing optimistic rendering to begin with. However, we should instrument on slow connections and identify areas where the user expects an immediate response but instead has to wait. We may want to introduce optimistic rendering only for common actions like thread/comment/reaction creation, and not across the board.